niimpy/location.py

```diff
@@ -196,36 +220,31 @@ def compute_distance_features(df):
         speed_max = np.nanmax(speeds)
         total_dist = sum(dists)
 
-        static_bins = speeds < 0.277
-        static_bins_index = np.arange(len(static_bins))[static_bins]
-        dists_matrix = np.zeros((len(static_bins_index), len(static_bins_index)))
-
-        lats_binned = df['double_latitude']
-        lons_binned = df['double_longitude']
-
-        for i, loc_index_i in enumerate(static_bins_index):
-            for j, loc_index_j in enumerate(static_bins_index):
-                lat_lon_i = lats_binned[loc_index_i], lons_binned[loc_index_i]
-                lat_lon_j = lats_binned[loc_index_j], lons_binned[loc_index_j]
-                dists_matrix[i][j] = geodesic(lat_lon_i, lat_lon_j).meters
-                dists_matrix[j][i] = dists_matrix[i][j]
-
-        from sklearn.cluster import DBSCAN
-        import collections
+        static_bins = speeds < speed_threshold
+        lats_static = lats[static_bins]
+        lons_static = lons[static_bins]
+        dists_matrix = distance_matrix(lats_static, lons_static)
 
         dbscan = DBSCAN(min_samples=5, eps=20, metric='precomputed')
         clusters = dbscan.fit_predict(dists_matrix)
         non_rare_clusters = clusters[clusters != -1]
         n_unique_sps = len(set(non_rare_clusters))
-        n_rare_sps = len(set(clusters)) - n_unique_sps
 
-        stay_times = collections.Counter(clusters).values()
+        counter = collections.Counter(clusters)
+        stay_times = counter.values()
         stay_times = np.sort(list(stay_times))[::-1]
-        n_bins = df.shape[0]
 
-        stay_perc_top1 = stay_times[0] / n_bins if len(stay_times) > 0 else 0
-        stay_perc_top2 = stay_times[1] / n_bins if len(stay_times) > 1 else 0
-        stay_perc_top3 = stay_times[2] / n_bins if len(stay_times) > 2 else 0
+        time_static = bin_width * sum(static_bins)
+        time_moving = bin_width * sum(~static_bins)
+        time_rare = bin_width * counter[-1]
+
+        n_transitions = sum(np.diff(clusters) != 0)
+
+        stay_top1 = stay_times[0] * bin_width if len(stay_times) > 0 else 0
+        stay_top2 = stay_times[1] * bin_width if len(stay_times) > 1 else 0
+        stay_top3 = stay_times[2] * bin_width if len(stay_times) > 2 else 0
+        stay_top4 = stay_times[3] * bin_width if len(stay_times) > 3 else 0
+        stay_top5 = stay_times[4] * bin_width if len(stay_times) > 4 else 0
```

Comment:
Is the amount of _stay_top_ related to the number of clusters? If yes then I think this code could be shortened a bit by just reporting the amount of found clusters instead of the top 5 stays.  # clusters can be included in the final dict, something like _n_cluster_.